#**Ask a question**
#Como recomendar uma cerveja para um determinado usuário baseando-se nos seus gostos?

#**Get the data**
#Faremos a seguir o upload do nosso data set de review de cervejas (obtido no site Kaggel, do usuário Sveen De Waard

In [2]:
import pandas as pd
import numpy as np
import sklearn
import matplotlib
import sqlite3






Importando os dados

In [3]:
df = pd.read_csv("beer_reviews.csv")
df.head()

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  
0            Sausa Weizen       5.0        47986  
1                Red Moon       6.2        48213  
2  Black Horse Black Beer       6.5        48215  
3              Sausa Pils       5.0        47969  
4           Cauldron DIPA       7.7        64883

In [4]:
df.describe()

brewery_id   review_time  review_overall  review_aroma  \
count  1.586614e+06  1.586614e+06    1.586614e+06  1.586614e+06   
mean   3.130099e+03  1.224089e+09    3.815581e+00  3.735636e+00   
std    5.578104e+03  7.654427e+07    7.206219e-01  6.976167e-01   
min    1.000000e+00  8.406720e+08    0.000000e+00  1.000000e+00   
25%    1.430000e+02  1.173224e+09    3.500000e+00  3.500000e+00   
50%    4.290000e+02  1.239203e+09    4.000000e+00  4.000000e+00   
75%    2.372000e+03  1.288568e+09    4.500000e+00  4.000000e+00   
max    2.800300e+04  1.326285e+09    5.000000e+00  5.000000e+00   

       review_appearance  review_palate  review_taste      beer_abv  \
count       1.586614e+06   1.586614e+06  1.586614e+06  1.518829e+06   
mean        3.841642e+00   3.743701e+00  3.792860e+00  7.042387e+00   
std         6.160928e-01   6.822184e-01  7.319696e-01  2.322526e+00   
min         0.000000e+00   1.000000e+00  1.000000e+00  1.000000e-02   
25%         3.500000e+00   3.500000e+00  3.500000e+00  5.200000e+00   
50%         4.000000e+00   4.000000e+00  4.000000e+00  6.500000e+00   
75%         4.000000e+00   4.000000e+00  4.500000e+00  8.500000e+00   
max         5.000000e+00   5.000000e+00  5.000000e+00  5.770000e+01   

        beer_beerid  
count  1.586614e+06  
mean   2.171279e+04  
std    2.181834e+04  
min    3.000000e+00  
25%    1.717000e+03  
50%    1.390600e+04  
75%    3.944100e+04  
max    7.731700e+04

Verificando se a coluna "review_overall" é calculada com base na MÉDIA das colunas "review_aroma", "review_appearance", "review_palate" e "review_taste"

In [5]:
values = []
for i in range(0, 1000):
    result = round((df.iloc[i]['review_aroma'] + df.iloc[i]['review_appearance'] + df.iloc[i]['review_palate'] + df.iloc[i]['review_taste'])/4)
    values.append((df.iloc[i]['review_overall'], result))

values

[(1.5, 2),
 (3.0, 3),
 (3.0, 3),
 (3.0, 3),
 (4.0, 4),
 (3.0, 3),
 (3.5, 4),
 (3.0, 3),
 (4.0, 4),
 (4.5, 4),
 (3.0, 2),
 (5.0, 4),
 (4.0, 4),
 (4.0, 3),
 (3.5, 4),
 (3.0, 3),
 (2.0, 2),
 (4.0, 4),
 (2.5, 2),
 (4.5, 4),
 (4.0, 4),
 (4.0, 4),
 (4.5, 4),
 (4.0, 4),
 (4.5, 4),
 (5.0, 4),
 (4.5, 4),
 (4.5, 4),
 (4.5, 4),
 (4.5, 4),
 (3.0, 2),
 (5.0, 4),
 (4.0, 4),
 (4.5, 4),
 (4.0, 4),
 (4.0, 4),
 (3.0, 3),
 (4.0, 4),
 (3.5, 4),
 (4.5, 4),
 (3.0, 2),
 (4.5, 4),
 (5.0, 4),
 (4.0, 4),
 (4.0, 4),
 (3.5, 4),
 (4.0, 4),
 (4.5, 4),
 (4.0, 4),
 (4.5, 4),
 (4.0, 4),
 (3.5, 4),
 (3.5, 4),
 (4.0, 4),
 (3.5, 4),
 (4.5, 4),
 (4.5, 4),
 (4.5, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (2.5, 3),
 (4.0, 4),
 (4.5, 5),
 (4.5, 4),
 (4.5, 4),
 (4.0, 4),
 (4.0, 4),
 (3.5, 4),
 (3.5, 4),
 (5.0, 4),
 (3.5, 4),
 (4.5, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (4.0, 4),
 (3.5, 4),
 (4.0, 3),
 (4.0, 4),
 (3.5, 4),
 (4.5, 4),
 (4.5, 4),
 (5.0, 4),
 (5.0, 4),
 (3.5, 4),
 (4.0, 4),

Parece que não!



Verificando qual cerveja é a mais avaliada no dataset

In [6]:
df['beer_name'].value_counts().idxmax()

'90 Minute IPA'

In [7]:
df[df['beer_name'] == '90 Minute IPA']

brewery_id          brewery_name  review_time  review_overall  \
1166284       10099  Dogfish Head Brewery   1266775164             4.0   
1166905       10099  Dogfish Head Brewery   1296628385             4.0   
1170249       10099  Dogfish Head Brewery   1250389086             4.0   
1170904       10099  Dogfish Head Brewery   1230224468             5.0   
1171046       10099  Dogfish Head Brewery   1229312994             4.0   
...             ...                   ...          ...             ...   
1176277       10099  Dogfish Head Brewery   1319407302             4.5   
1176278       10099  Dogfish Head Brewery   1319326828             4.0   
1176279       10099  Dogfish Head Brewery   1319211417             4.5   
1176280       10099  Dogfish Head Brewery   1319161135             4.5   
1176281       10099  Dogfish Head Brewery   1319143049             4.0   

         review_aroma  review_appearance review_profilename  \
1166284           4.5                4.5            DmanGTR   
1166905           4.0                4.0         nickadams2   
1170249           4.0                4.0           kegger22   
1170904           5.0                5.0       fairway31533   
1171046           4.5                4.0               Bung   
...               ...                ...                ...   
1176277           4.0                4.5         kennylopez   
1176278           4.0                4.0          firedog23   
1176279           4.5                4.5             heysuz   
1176280           4.5                5.0          match1112   
1176281           4.0                4.0           tbeckett   

                             beer_style  review_palate  review_taste  \
1166284  American Double / Imperial IPA            4.5           4.5   
1166905  American Double / Imperial IPA            3.5           4.0   
1170249  American Double / Imperial IPA            4.0           4.5   
1170904  American Double / Imperial IPA            5.0           5.0   
1171046  American Double / Imperial IPA            4.0           4.0   
...                                 ...            ...           ...   
1176277  American Double / Imperial IPA            4.5           4.0   
1176278  American Double / Imperial IPA            4.5           4.0   
1176279  American Double / Imperial IPA            4.5           4.5   
1176280  American Double / Imperial IPA            4.0           4.5   
1176281  American Double / Imperial IPA            4.0           4.0   

             beer_name  beer_abv  beer_beerid  
1166284  90 Minute IPA       9.0         2093  
1166905  90 Minute IPA       9.0         2093  
1170249  90 Minute IPA       9.0         2093  
1170904  90 Minute IPA       9.0         2093  
1171046  90 Minute IPA       9.0         2093  
...                ...       ...          ...  
1176277  90 Minute IPA       9.0         2093  
1176278  90 Minute IPA       9.0         2093  
1176279  90 Minute IPA       9.0         2093  
1176280  90 Minute IPA       9.0         2093  
1176281  90 Minute IPA       9.0         2093  

[3290 rows x 13 columns]

Verificando a segunda cerveja mais avaliada

In [9]:
df['beer_name'].value_counts()

beer_name
90 Minute IPA                          3290
India Pale Ale                         3130
Old Rasputin Russian Imperial Stout    3111
Sierra Nevada Celebration Ale          3000
Two Hearted Ale                        2728
                                       ... 
Titanbraü Pils                            1
Titanbraü Weizen                          1
Titanbräu Ale                             1
Texas Brunette                            1
Sausa Weizen                              1
Name: count, Length: 56857, dtype: int64

In [10]:
df[df['beer_name'] == 'Indian Pale Ale']['review_overall']

brewery_id                     brewery_name  review_time  \
633356        10089        Redoak Boutique Beer Café   1314092366   
633357        10089        Redoak Boutique Beer Café   1307682489   
633358        10089        Redoak Boutique Beer Café   1304036377   
633359        10089        Redoak Boutique Beer Café   1282622103   
633360        10089        Redoak Boutique Beer Café   1263020024   
633361        10089        Redoak Boutique Beer Café   1193897311   
633362        10089        Redoak Boutique Beer Café   1187597117   
633363        10089        Redoak Boutique Beer Café   1181128736   
633364        10089        Redoak Boutique Beer Café   1162182662   
633365        10089        Redoak Boutique Beer Café   1156546829   
805602        10960            Ramapo Valley Brewery   1161451058   
805603        10960            Ramapo Valley Brewery   1137970519   
805604        10960            Ramapo Valley Brewery   1132777286   
1024158        1863    Southend Brewery & Smokehouse   1181529130   
1024177        1863    Southend Brewery & Smokehouse   1095468273   
1230911        9243       Brewery At Lake Tahoe, The   1313775093   
1230912        9243       Brewery At Lake Tahoe, The   1285162192   
1230913        9243       Brewery At Lake Tahoe, The   1228521940   
1230914        9243       Brewery At Lake Tahoe, The   1205820961   
1230915        9243       Brewery At Lake Tahoe, The   1125106495   
1472196        1086  Cleveland ChopHouse And Brewery   1178043934   
1484161        7216         Adirondack Pub & Brewery   1322954778   
1484162        7216         Adirondack Pub & Brewery   1286232177   
1484163        7216         Adirondack Pub & Brewery   1252460347   

         review_overall  review_aroma  review_appearance review_profilename  \
633356              3.5           3.5                4.0             Stew41   
633357              4.0           4.0                4.0          heygeebee   
633358              4.5           4.0                4.0          heygeebee   
633359              4.5           4.0                4.5    laituegonflable   
633360              4.0           4.5                4.5              juju7   
633361              4.0           4.0                4.0     lacqueredmouse   
633362              4.5           4.0                4.5         jarmby1711   
633363              4.0           4.5                4.0          brendan13   
633364              3.0           3.0                3.0           gangrel7   
633365              4.5           3.5                3.5           vancurly   
805602              3.5           3.0                4.0           ppoitras   
805603              1.5           4.0                4.0        johnniemarg   
805604              3.0           4.0                4.0         NJpadreFan   
1024158             3.5           3.5                3.0          jackorain   
1024177             3.0           2.0                3.0               Boto   
1230911             4.0           3.5                4.0         Haybeerman   
1230912             4.5           3.5                4.0           dickroop   
1230913             3.5           3.5                3.5          benclimbs   
1230914             3.5           3.5                3.5        brandoneast   
1230915             3.5           3.0                4.0                KoG   
1472196             4.0           3.5                4.0         rhoadsrage   
1484161             3.5           3.5                3.5          UDbeernut   
1484162             4.0           4.5                4.5         TangTKE402   
1484163             4.0           4.0                4.0            plaid75   

                           beer_style  review_palate  review_taste  \
633356   English India Pale Ale (IPA)            3.0           3.5   
633357   English India Pale Ale (IPA)            3.0           4.5   
633358   English India Pale Ale (IPA)            3.5           4.5   
633359   English India Pale

Atribuindo um id único a cada avaliador da coluna "review_profilename"

In [22]:
# Suponha que a coluna com os nomes dos avaliadores é chamada 'Nome_Avaliador'
avaliadores = df['review_profilename'].unique()
avaliador_id_map = {nome: i for i, nome in enumerate(avaliadores, 1)}

df['ID_Avaliador'] = df['review_profilename'].map(avaliador_id_map)

df



brewery_id                 brewery_name  review_time  review_overall  \
0             10325              Vecchio Birraio   1234817823             1.5   
1             10325              Vecchio Birraio   1235915097             3.0   
2             10325              Vecchio Birraio   1235916604             3.0   
3             10325              Vecchio Birraio   1234725145             3.0   
4              1075      Caldera Brewing Company   1293735206             4.0   
...             ...                          ...          ...             ...   
1586609       14359  The Defiant Brewing Company   1162684892             5.0   
1586610       14359  The Defiant Brewing Company   1161048566             4.0   
1586611       14359  The Defiant Brewing Company   1160702513             4.5   
1586612       14359  The Defiant Brewing Company   1160023044             4.0   
1586613       14359  The Defiant Brewing Company   1160005319             5.0   

         review_aroma  review_appearance review_profilename  \
0                 2.0                2.5            stcules   
1                 2.5                3.0            stcules   
2                 2.5                3.0            stcules   
3                 3.0                3.5            stcules   
4                 4.5                4.0     johnmichaelsen   
...               ...                ...                ...   
1586609           4.0                3.5         maddogruss   
1586610           5.0                2.5          yelterdow   
1586611           3.5                3.0           TongoRad   
1586612           4.5                4.5           dherling   
1586613           4.5                4.5               cbl2   

                             beer_style  review_palate  review_taste  \
0                            Hefeweizen            1.5           1.5   
1                    English Strong Ale            3.0           3.0   
2                Foreign / Export Stout            3.0           3.0   
3                       German Pilsener            2.5           3.0   
4        American Double / Imperial IPA            4.0           4.5   
...                                 ...            ...           ...   
1586609                     Pumpkin Ale            4.0           4.0   
1586610                     Pumpkin Ale            2.0           4.0   
1586611                     Pumpkin Ale            3.5           4.0   
1586612                     Pumpkin Ale            4.5           4.5   
1586613                     Pumpkin Ale            4.5           4.5   

                      beer_name  beer_abv  beer_beerid  ID_Avaliador  
0                  Sausa Weizen       5.0        47986             1  
1                      Red Moon       6.2        48213             1  
2        Black Horse Black Beer       6.5        48215             1  
3                    Sausa Pils       5.0        47969             1  
4                 Cauldron DIPA       7.7        64883             2  
...                         ...       ...          ...           ...  
1586609      The Horseman's Ale       5.2        33061          5888  
1586610      The Horseman's Ale       5.2        33061           707  
1586611      The Horseman's Ale       5.2        33061          7047  
1586612      The Horseman's Ale       5.2        33061          3980  
1586613      The Horseman's Ale       5.2        33061          2575  

[1586614 rows x 14 columns]

In [26]:
df.columns

Index(['brewery_id', 'brewery_name', 'review_time', 'review_overall',
       'review_aroma', 'review_appearance', 'review_profilename', 'beer_style',
       'review_palate', 'review_taste', 'beer_name', 'beer_abv', 'beer_beerid',
       'ID_Avaliador'],
      dtype='object')

Descartando algumas colunas que não serão úteis

In [27]:
df = df[['beer_beerid', 'beer_name', 'beer_abv', 'beer_style', 'ID_Avaliador', 'review_overall']]
df

beer_beerid               beer_name  beer_abv  \
0              47986            Sausa Weizen       5.0   
1              48213                Red Moon       6.2   
2              48215  Black Horse Black Beer       6.5   
3              47969              Sausa Pils       5.0   
4              64883           Cauldron DIPA       7.7   
...              ...                     ...       ...   
1586609        33061      The Horseman's Ale       5.2   
1586610        33061      The Horseman's Ale       5.2   
1586611        33061      The Horseman's Ale       5.2   
1586612        33061      The Horseman's Ale       5.2   
1586613        33061      The Horseman's Ale       5.2   

                             beer_style  ID_Avaliador  review_overall  
0                            Hefeweizen             1             1.5  
1                    English Strong Ale             1             3.0  
2                Foreign / Export Stout             1             3.0  
3                       German Pilsener             1             3.0  
4        American Double / Imperial IPA             2             4.0  
...                                 ...           ...             ...  
1586609                     Pumpkin Ale          5888             5.0  
1586610                     Pumpkin Ale           707             4.0  
1586611                     Pumpkin Ale          7047             4.5  
1586612                     Pumpkin Ale          3980             4.0  
1586613                     Pumpkin Ale          2575             5.0  

[1586614 rows x 6 columns]

Alterando o nome das colunas

In [28]:
columns = {'beer_beerid':'id_cerveja', 'beer_name':'nome_cerveja', 'beer_abv':'teor_alcoolico', 'beer_style':'tipo_cerveja', 'review_overall':'nota'}
df.rename(columns = columns, inplace=True)
df

C:\Users\marce\AppData\Local\Temp\ipykernel_6020\1851634588.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns = columns, inplace=True)


id_cerveja            nome_cerveja  teor_alcoolico  \
0             47986            Sausa Weizen             5.0   
1             48213                Red Moon             6.2   
2             48215  Black Horse Black Beer             6.5   
3             47969              Sausa Pils             5.0   
4             64883           Cauldron DIPA             7.7   
...             ...                     ...             ...   
1586609       33061      The Horseman's Ale             5.2   
1586610       33061      The Horseman's Ale             5.2   
1586611       33061      The Horseman's Ale             5.2   
1586612       33061      The Horseman's Ale             5.2   
1586613       33061      The Horseman's Ale             5.2   

                           tipo_cerveja  ID_Avaliador  nota  
0                            Hefeweizen             1   1.5  
1                    English Strong Ale             1   3.0  
2                Foreign / Export Stout             1   3.0  
3                       German Pilsener             1   3.0  
4        American Double / Imperial IPA             2   4.0  
...                                 ...           ...   ...  
1586609                     Pumpkin Ale          5888   5.0  
1586610                     Pumpkin Ale           707   4.0  
1586611                     Pumpkin Ale          7047   4.5  
1586612                     Pumpkin Ale          3980   4.0  
1586613                     Pumpkin Ale          2575   5.0  

[1586614 rows x 6 columns]

In [29]:
df['ID_Avaliador'].value_counts()

ID_Avaliador
43       5817
129      4661
64       4617
155      3518
199      3497
         ... 
23971       1
23974       1
23975       1
23978       1
33388       1
Name: count, Length: 33388, dtype: int64

Quantos usuários avaliaram pelo menos 50 cervejas?


In [64]:
y = df['ID_Avaliador'].value_counts() >= 50
y[y == True]

ID_Avaliador
43      True
129     True
64      True
155     True
199     True
        ... 
5101    True
7937    True
6882    True
5378    True
3936    True
Name: count, Length: 4787, dtype: bool

De 33388 usuários, 4732 avaliaram pelo menos 50 cervejas! Queremos utilizar apenas esses usuários.

In [67]:
w = y[y].index.tolist() #Lista dos ID's dos avaliadores que avaliaram pelo menos 50 cervejas
av_50 = df[df['ID_Avaliador'].isin(w)]
av_50

id_cerveja            nome_cerveja  teor_alcoolico  \
0             47986            Sausa Weizen             5.0   
1             48213                Red Moon             6.2   
2             48215  Black Horse Black Beer             6.5   
3             47969              Sausa Pils             5.0   
4             64883           Cauldron DIPA             7.7   
...             ...                     ...             ...   
1586609       33061      The Horseman's Ale             5.2   
1586610       33061      The Horseman's Ale             5.2   
1586611       33061      The Horseman's Ale             5.2   
1586612       33061      The Horseman's Ale             5.2   
1586613       33061      The Horseman's Ale             5.2   

                           tipo_cerveja  ID_Avaliador  nota  
0                            Hefeweizen             1   1.5  
1                    English Strong Ale             1   3.0  
2                Foreign / Export Stout             1   3.0  
3                       German Pilsener             1   3.0  
4        American Double / Imperial IPA             2   4.0  
...                                 ...           ...   ...  
1586609                     Pumpkin Ale          5888   5.0  
1586610                     Pumpkin Ale           707   4.0  
1586611                     Pumpkin Ale          7047   4.5  
1586612                     Pumpkin Ale          3980   4.0  
1586613                     Pumpkin Ale          2575   5.0  

[1396114 rows x 6 columns]